src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs

```diff
@@ -752,11 +759,41 @@ public SymbolComparer(MatchSymbols matcher)
                     _matcher = matcher;
                 }
 
-                public bool Equals(Symbol x, Symbol y)
+                public bool Equals(Symbol left, Symbol right)
                 {
-                    var other = _matcher.Visit(x);
-                    Debug.Assert((object)other != null);
-                    return other == y;
+                    if (left.Kind != right.Kind)
+                    {
+                        return false;
+                    }
+
+                    var anonymousLeft = left as AnonymousTypeManager.AnonymousTypePublicSymbol;
+                    if ((object)anonymousLeft != null)
+                    {
+                        left = AnonymousTypeManager.TranslateAnonymousTypeSymbol(anonymousLeft);
+                    }
+
+                    var anonymousRight = right as AnonymousTypeManager.AnonymousTypePublicSymbol;
+                    if ((object)anonymousRight != null)
+                    {
+                        right = AnonymousTypeManager.TranslateAnonymousTypeSymbol(anonymousRight);
+                    }
+
+                    if ((object)anonymousLeft != null || (object)anonymousRight != null)
+                    {
+                        var namedLeft = (NamedTypeSymbol)left;
+                        var namedRight = (NamedTypeSymbol)right;
+
+                        // Map the template - this involves comparing the structure of the anonymous type,
+                        // so we don't compare equal types with different field names.
+                        if (_matcher.Visit(namedLeft.ConstructedFrom)?.Equals(namedRight.ConstructedFrom) != true)
```

Comment:
Is it possible to get into a cycle here (ConstructedFrom can return the same symbol back)?